DOCSP-34813 Data Verification (#241)
mongodb/docs-cluster-to-cluster-sync: source/reference/verification.txt

====================
Verify Data Transfer
====================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Following a sync, verify the successful transfer of your data before you switch
your application from using the source cluster to the destination cluster. You
can verify your data transfer using document counts, hash comparison, document
comparison, or the Migration Verifier.

Use Cases
---------

You should verify your data after every sync. This is important in
cases where you plan to move your application load from the source to the
destination cluster.

> Verification methods include the comparison of

Reviewer: Maybe consider using a bulleted list for these methods.

Author: Fixed

:ref:`Document Counts <c2c-verify-doc-counts>`,
comparison of :ref:`Hash Comparison <c2c-verify-hash-comp>`,
scripting :ref:`Document Comparisons <c2c-verify-doc-comp>`,
and the :ref:`Migration Verifier <c2c-verify-migration-verifier>`.
The specific method you use to verify your data depends on your application
workload and the complexity of the data.

.. _c2c-verify-method:

> Tasks

Reviewer: Thanks for this refactor. I think this is a big improvement to the presentation.

====================
Verify Data Transfer
====================

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

> Following a sync, verify the successful transfer of your data before you switch

Reviewer 2: I think we should make it clear that we mean once mongosync has fully committed.

Author: Fixed

> your application from using the source cluster to the destination cluster. You

Reviewer 2: ```suggestion
your application to using the destination cluster. You
```

Reviewer 2: I don't want to confuse users into thinking that writes on the source cluster should be continued after commit before verification

Author: Fixed

cluster. You can verify your data transfer using document counts, hash
comparison, document comparison, or the Migration Verifier.

Use Cases
---------

You should verify your data after every sync. This is important in
cases where you plan to move your application load from the source to the
destination cluster.

Verification methods:

- :ref:`c2c-verify-with-doc-counts`

- :ref:`c2c-verify-with-hash-comp`

- :ref:`c2c-verify-with-doc-comp`

- :ref:`c2c-verify-with-verifier`

The specific method you use to verify your data depends on your application
workload and the complexity of the data.

.. _c2c-verify-method:

Tasks
-------

.. _c2c-verify-with-doc-counts:

Document Counts
~~~~~~~~~~~~~~~

The most basic method of verifying sync is to compare the number of documents
in each synced collection on the source cluster to the number on the
destination cluster.

> While this method is fast and nonblocking, it only verifies a successful sync when

Author: Fixed

run against clusters with insert-only workloads.

For more information, see :ref:`c2c-verify-doc-counts`.

.. _c2c-verify-with-hash-comp:

Hash Comparison
~~~~~~~~~~~~~~~

You can verify sync by comparing MD5 hashes of collections synced from the
source cluster to the destination cluster.

While hash comparison ensures that the destination cluster has received all
changes from the source, the :dbcommand:`dbHash` command locks the cluster,
preventing additional writes until it completes.

.. note::

   Hash comparison is not possible with sharded clusters. It also does not
   work for standalone servers and replica sets that use MongoDB 4.4 or earlier
   releases, since the document field order can vary.

For more information, see :ref:`c2c-verify-hash-comp`.

.. _c2c-verify-with-doc-comp:

Document Comparison
~~~~~~~~~~~~~~~~~~~

You can verify sync by comparing documents on the source and destination
clusters. Write a script that queries collections on the source cluster and
then checks that the document exists with the same values on the destination
cluster.

.. _c2c-verify-with-verifier:

> Migration Verifier

Reviewer 3: I think we're missing one of the key selling points of the migration verifier vs document comparison. The verifier can verify changes during the migration, which means the final verification during cutover can be quick.

Author: Fixed

.. _c2c-verification:

====================
Verify Data Transfer
====================

.. default-domain:: mongodb

.. facet::
   :name: genre
   :values: reference

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

When :program:`mongosync` has fully committed, verify the successful transfer
of your data before you switch your application to using the destination
cluster. You can verify your data transfer using document counts, hash
comparison, document comparison, or the Migration Verifier.

Use Cases
---------

You should verify your data after every sync. This is important in
cases where you plan to move your application load from the source to the
destination cluster.

Verification methods:

- :ref:`c2c-verify-with-doc-counts`

- :ref:`c2c-verify-with-hash-comp`

- :ref:`c2c-verify-with-doc-comp`

- :ref:`c2c-verify-with-verifier`

The specific method you use to verify your data depends on your application
workload and the complexity of the data.

.. _c2c-verify-method:

Tasks
-------

.. _c2c-verify-with-doc-counts:

Document Counts
~~~~~~~~~~~~~~~

> The most basic method of verifying sync is to compare the number of documents

Reviewer 2: ```suggestion
The most basic method of verification is to compare the number of documents
```

changes from the source, the :dbcommand:`dbHash` command locks the cluster,
preventing additional writes until it completes.

.. note::

   Hash comparison is not possible with sharded clusters. It also does not
   work for standalone servers and replica sets that use MongoDB 4.4 or earlier
   releases, since the document field order can vary.

For more information, see :ref:`c2c-verify-hash-comp`.

.. _c2c-verify-with-doc-comp:

Document Comparison
~~~~~~~~~~~~~~~~~~~

You can verify sync by comparing documents on the source and destination
clusters. Write a script that queries collections on the source cluster and
then checks that the document exists with the same values on the destination
cluster.

.. _c2c-verify-with-verifier:

Migration Verifier
~~~~~~~~~~~~~~~~~~

Migration Verifier connects to the source and destination clusters and performs
a series of verification checks, comparing documents, views, and indexes to
confirm the sync was successful.

.. important::

   Migration Verifier is an experimental and unsupported tool.

   For installation instructions, see
   `GitHub <https://github.com/mongodb-labs/migration-verifier>`__.

> Unlike other verification methods, Migration Verifier can run concurrent

Reviewer 2: ```suggestion
Unlike other verification methods, Migration Verifier can run concurrently
```

> ####################

Author: Fixed

Verify Data Transfer
####################

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Following a sync, verify the successful transfer of your data before you switch
your application from using the source cluster to the destination cluster. You
can verify your data transfer using document counts, hash comparison, document
comparison, or the Migration Verifier.

Use Cases
=========

You should verify your data after every sync. This is important in
cases where you plan to move your application load from the source to the
destination cluster.

Verification methods include the comparison of
:ref:`document counts <c2c-verify-doc-counts>`,
comparison of :ref:`database hashes <c2c-verify-hash-comp>`,
scripting :ref:`document comparisons <c2c-verify-doc-comp>`,
and the :ref:`Migration Verifier <c2c-verify-migration-verifier>`.
The specific method you use to verify your data depends on your application
workload and the complexity of the data.

Details
=======

.. _c2c-verify-doc-counts:

> Document Counts

Reviewer: I'm a little confused as to whether this is a task page or a concept page. Looking at our IA templates, the Use Cases and Details sections align with the [concept](https://www.mongodb.com/docs/meta/style-guide/information-types/prototypes/concept/) page, but we're also kind of showing users how to accomplish the task here.

If we're using Use Cases and Details sections and treating this as a conceptual page, I'd maybe expect to see the specific methods appear in a Task section, with each having their own page.

Author: Refactored to split the examples out into tasks.

---------------

> The most basic method of verifying sync is by comparing the number of documents

Reviewer: I think we can remove "by" from this sentence.

```suggestion
The most basic method of verifying sync is comparing the number of documents
```

Author: Fixed

in each synced collection on the source cluster to the number on the
destination cluster.

Compare the results on each cluster:

- If the document count is different,  not all changes synced to the
  destination cluster.

- If the document count is the same, the destination cluster has the same
  number of documents. This doesn’t guarantee a successful sync of all changes

>   to those documents, except in cases where you have an insert-only workload.

Reviewer: I want to make sure I'm understanding this sentence correctly:

> This doesn’t guarantee a successful sync of all changes
  to those documents, except in cases where you have an insert-only workload.

Is this saying that verifying that the document counts are the same does not ensure that modifications to the data (as in update operations) are accounted for in the sync? As in, the document counts might be the same but that doesn't necessarily mean that the correct data was synced?

Author: Reworked for clarity.

To verify that the source and destination cluster have the same number of
documents, run the :method:`~db.collection.countDocuments` method for each synced
collection on each cluster and ensure that they have the same number of

> documents.

Author: Fixed

.. io-code-block::

   .. input::
      :language: javascript

      db.us_accounts.countDocuments()

   .. output::
      :language: javascript

      345

.. _c2c-verify-hash-comp:

Hash Comparison
---------------

You can verify sync by comparing MD5 hashes of collections synced from the

> source cluster to the destination cluster. While hash comparison does ensure

Reviewer: ```suggestion
source cluster to the destination cluster. While hash comparison ensures
```

Author: Fixed

that the destination cluster has received all changes from the source, the
:dbcommand:`dbHash` command locks the cluster, preventing additional writes

> until its operations complete.

Reviewer: ```suggestion
until it completes.
```

Author: Fixed

.. note::

   Hash comparison is not possible with sharded clusters. It also does not
   work for standalone servers and replica sets that use MongoDB 4.4 or earlier

>    releases, since the field order can vary.

Reviewer: Is the "field order" here referring to document field order?

Author: Fixed to "document field order"

> To verify that the source and destination cluster have the same hashes, execute

Reviewer: We use "run" for the previous section `countDocuments()`.

```suggestion
To verify that the source and destination cluster have the same hashes, run
```

Author: Fixed

the :dbcommand:`dbHash` command for the synced collections on each cluster and
compare the hash value in the ``md5`` field of the return documents.

.. io-code-block::

   .. input::
      :language: javascript

      db.runCommand({
         dbhash: 1,
         collections: [
            "accounts.us_accounts",
            "accounts.eu_accounts",
            ...
         ]}).md5

   .. output::
      :language: text

>       d41d8cd98f00b204e9800998ecf8427e

Reviewer: [Question]

Looking at the `dbHash` docs, it looks like we return a hash for each collection specified in the command. Should we include multiple hashes in the example output to make that clear?

Author: @jeff-allen-mongo, I may need to clean up the return content a bit.  The `md5` field returns an aggregate value for the hashed collections, which I imagine makes it easier to check for matches.

Reviewer: That makes sense to me 👍

.. _c2c-verify-doc-comp:

Document Comparison
-------------------

You can verify sync by comparing documents on the source and destination
clusters. Write a script that queries collections on the source cluster and
then checks that the document exists with the same values on the destination
cluster.

.. _c2c-verify-migration-verifier:

Migration Verifier
------------------

> Migration Verifier connects to the source and destination clusters and performs

Reviewer: [question]

I'm assuming that the verifier comes packaged with the Relational Migrator install, and doesn't require anything additional to run. If that's the case, maybe that point is worth clarifying?

Author: @jeff-allen-mongo, I can't speak to Relational Migrator since I'm not familiar with it.  The link the C2C team provided goes to a separate GitHub project, which is compiled through Go tooling.  I link to it in the "For more information" bit at the bottom, but can  add some installation instructions if you think it's needed.

Reviewer: I don't think we need full installation instructions, but it might help to move the GitHub link higher up on the page and say that users should go there to learn how to install the migrator. What do you think?

Author: Added a link to the admonition calling out installation information.

a series of verification checks, comparing documents, views, and indexes to
confirm the sync was successful.

.. important::

   Migration Verifier is an experimental and unsupported tool.

To use the Migration Verifier,

#. Start Migration Verifier:

   .. code-block:: bash

      migration-verifier --verifyAll \
         --srcURI example.net:27020 \
         --destURI example.net:27021 \
         --metaURI example.net:27017

#. To start the verification checks, use cURL to send the ``check`` command to
   Migration Verifier:

   .. code-block:: bash

      curl -H "Content-Type: application/json" \
          -X POST -d '{}' http://127.0.0.1:27020/api/v1/check

#. When ``mongosync`` is complete, use cURL to send the ``writeOff`` command to
   Migration Verifier:

   .. code-block:: bash

      curl -H "Content-Type: application/json" \

>           -X POST -d '{}' http://127.0.0.1:27020/api/v1/writesOff

Reviewer: The step says `writeOff` but the curl command says `writesOff`. Which is correct?

Author: Fixed

#. To view the results, use cURL to send the ``progress`` command to Migration
   Verifier:

   .. io-code-block::

      .. input::
         :language: bash

         curl -H "Content-Type: application/json" \
            -X GET http://127.0.0.1:27020/api/v1/progress

      .. output::
         :language: json

         {
            "progress": {
               "phase": "idle",
               "error": null,
               "verificationStatus": {
                  "totalTasks": 1,
                  "addedTasks": 0,
                  "processingTasks": 0,
                  "failedTasks": 1,
                  "completedTasks": 0,
                  "metadataMismatchTasks": 0,
                  "recheckTasks": 0
               }
            }
         }

When the ``phase`` field returns ``idle``, the verification checks are

> complete.

Reviewer: Can we be specific about what the user should be checking in the output to confirm whether or not the sync was successful?

Author: Fixed
